src/rpc/AppRpcProvider.ts

```diff
@@ -0,0 +1,156 @@
+          lastEvaluated: 0,
+        },
+      })
+    })
+
+    this.evaluationInterval = evaluationInterval
+  }
+
+  /**
+   * Perform a JSON-RPC request.
+   *
+   * method - The JSON-RPC method name.
+   * params - The parameters for the JSON-RPC method.
+   * Returns a Promise that resolves with the result of the JSON-RPC method.
+   * Throws an error if all providers fail to perform the operation.
+   */
+  async perform(method: string, params: { [name: string]: any }): Promise<any> {
+    // Periodically evaluate all providers
+    const currentTime = now()
+    this.providerEvaluations.forEach((providerEval) => {
+      if (currentTime - providerEval.performance.lastEvaluated >= this.evaluationInterval) {
+        this.evaluateProvider(providerEval)
+      }
+    })
+
+    // Sort providers by performance score (latency and failure rate)
+    const sortedEvaluations = AppRpcProvider.sortProviders(this.providerEvaluations.slice())
+
+    // Always broadcast "sendTransaction" to all backends
+    if (method === 'sendTransaction') {
+      const results: Array<string | Error> = await Promise.all(
+        sortedEvaluations.map(({ provider }) => {
+          return provider.sendTransaction(params.signedTransaction).then(
+            (result) => result.hash,
+            (error) => error
+          )
+        })
+      )
+
+      // Any success is good enough
+      for (let i = 0; i < results.length; i++) {
+        if (typeof results[i] === 'string') return results[i]
+      }
+
+      // They were all an error; pick the first error
+      throw results[0]
+    } else {
+      for (const { provider, performance } of sortedEvaluations) {
+        try {
+          return provider.perform(method, params)
+        } catch (error) {
+          performance.failureRate++ // Increment failure rate
+        }
+      }
+      throw new Error('All providers failed to perform the operation.')
+    }
+  }
+
+  /**
+   * Evaluates the performance of a provider. Updates latency and failure rate metrics.
+   *
+   * config - The provider evaluation configuration.
+   * Returns a Promise that resolves when the evaluation is complete.
+   */
+  async evaluateProvider(config: FallbackProviderEvaluation): Promise<void> {
+    const startTime = now()
+    try {
+      await config.provider.getBlockNumber()
+      const latency = now() - startTime
+      config.performance.latency = latency
+      config.performance.failureRate = 0 // Reset failure rate on successful request
+    } catch (error) {
+      config.performance.failureRate += 1 // Increase failure rate on failed request
+    }
+    config.performance.lastEvaluated = now()
+  }
+
+  static sortProviders(providerEvaluations: FallbackProviderEvaluation[]) {
+    return providerEvaluations.sort((a, b) => {
+      const scoreA = a.performance.latency * (a.performance.failureRate + 1)
+      const scoreB = b.performance.latency * (b.performance.failureRate + 1)
+      return scoreA - scoreB
+    })
+  }
```

Comment:
**example behavior**
Provider Name | Latency (ms) | Failure Rate | Score
-- | -- | -- | --
Provider B | 20 | 0 | 20
Provider A | 50 | 2 | 150
Provider E | 80 | 1 | 160
Provider C | 100 | 1 | 200
Provider D | 40 | 4 | 200

Reply:
Isn't it possible for something which fails quickly to rank before something which succeeds slowly?

Reply:
yes, but its failure rate would climb rapidly and all failed requests would be retried automatically at lower ranked venues immediately, so the ux effect is minimal imo. I'm fine with making tweaks to the algo if you have any in mind.